core/arch/arm/kernel/user_ta.c

```diff
@@ -692,6 +694,50 @@ static TEE_Result add_deps(struct user_ta_ctx *utc __unused,
 
 #endif
 
+#ifdef CFG_TA_ASLR
+static size_t aslr_offset(size_t min, size_t max)
+{
+	uint32_t rnd32 = 0;
+	size_t rnd = 0;
+
+	assert(min <= max);
+	if (max > min) {
+		if (crypto_rng_read(&rnd32, sizeof(rnd32))) {
+			DMSG("Random read failed");
+			return min;
+		}
+		rnd = rnd32 % (max - min);
+	}
+	return (min + rnd) * CORE_MMU_USER_CODE_SIZE;
+}
+
+static vaddr_t get_stack_va_hint(struct user_ta_ctx *utc)
+{
+	struct vm_region *r;
+	vaddr_t base = 0;
+
+	r = TAILQ_LAST(&utc->vm_info->regions, vm_region_head);
+	if (r) {
+		/*
+		 * Adding an empty page to separate TA mappings from already
+		 * present mappings with TEE_MATTR_PERMANENT to satisfy
+		 * select_va_in_range()
+		 */
+		base = r->va + r->size + CORE_MMU_USER_CODE_SIZE;
+	} else {
+		core_mmu_get_user_va_range(&base, NULL);
+	}
+
+	return base + aslr_offset(CFG_TA_ASLR_MIN_OFFSET_PAGES,
+				  CFG_TA_ASLR_MAX_OFFSET_PAGES);
+}
+#else
+static vaddr_t get_stack_va_hint(struct user_ta_ctx *utc __unused)
+{
+	return 0;
```

Comment:
I wonder why not `{ core_mmu_get_user_va_range(&base, NULL); return base; }` here?
Or why not `base = 0;` at line 728?
`

Reply:
> I wonder why not { core_mmu_get_user_va_range(&base, NULL); return base; } here?

It would achieve the same goal (selecting the first available VA), but Jens preferred I use `0` (see https://github.com/OP-TEE/optee_os/pull/2715#discussion_r243549926). You can't `return 0;` for simplicity ;-)

> Or why not `base = 0;` at line 728?

That would be incorrect, because `get_stack_va_hint()` would return `0 + aslr_offset(...);` which means nothing.